nifi-nar-bundles/nifi-riemann-bundle/nifi-riemann-processors/src/main/java/org/apache/nifi/processors/riemann/PutRiemann.java

```diff
@@ -0,0 +1,376 @@
+      batchSize = context.getProperty(BATCH_SIZE).asInteger();
+    }
+    if (riemannClient == null || !riemannClient.isConnected()) {
+      transport = Transport.valueOf(context.getProperty(TRANSPORT_PROTOCOL).getValue());
+      String host = context.getProperty(RIEMANN_HOST).getValue().trim();
+      int port = context.getProperty(RIEMANN_PORT).asInteger();
+      writeTimeout = context.getProperty(TIMEOUT).asLong();
+      RiemannClient client = null;
+      try {
+        switch (transport) {
+          case TCP:
+            client = RiemannClient.tcp(host, port);
+            break;
+          case UDP:
+            client = RiemannClient.udp(host, port);
+            break;
+        }
+        client.connect();
+        riemannClient = client;
+      } catch (IOException e) {
+        context.yield();
+        throw new ProcessException(String.format("Unable to connect to Riemann [%s:%d] (%s)\n%s", host, port, transport, e.getMessage()));
+      }
+    }
+
+    if (customAttributes.size() == 0) {
+      for (Map.Entry<PropertyDescriptor, String> property : context.getProperties().entrySet()) {
+        // only custom defined properties
+        if (!getSupportedPropertyDescriptors().contains(property.getKey())) {
+          customAttributes.add(property.getKey());
+        }
+      }
+    }
+  }
+
+
+  @Override
+  public void onTrigger(ProcessContext context, ProcessSession session) throws ProcessException {
+    // Check if the client is currently connected, as a previous trigger could have detected a failure
+    // in the connection.
+    if (riemannClient == null || !riemannClient.isConnected()) {
+      // clean up the client and attempt to re-initialize the processor
+      cleanUpClient();
+      onScheduled(context);
+    }
+
+    List<FlowFile> incomingFlowFiles = session.get(batchSize);
+    List<FlowFile> successfulFlowFiles = new ArrayList<>(incomingFlowFiles.size());
+    List<Event> eventsQueue = new ArrayList<>(incomingFlowFiles.size());
+    for (FlowFile flowFile : incomingFlowFiles) {
+      try {
+        eventsQueue.add(FlowFileToEvent.fromAttributes(context, customAttributes, flowFile));
+        successfulFlowFiles.add(flowFile);
+      } catch (NumberFormatException e) {
+        getLogger().warn(e.getMessage());
+        session.transfer(flowFile, REL_FAILURE);
+      }
+    }
+    try {
+      if (transport == Transport.TCP) {
+        Proto.Msg returnMessage = riemannClient.sendEvents(eventsQueue).deref(writeTimeout, TimeUnit.MILLISECONDS);
+        if (returnMessage == null) {
+          context.yield();
+          throw new ProcessException("Timed out writing to Riemann!");
+        }
+      } else {
+        riemannClient.sendEvents(eventsQueue);
+      }
+      riemannClient.flush();
+    } catch (Exception e) {
+      context.yield();
+      throw new ProcessException("Failed writing to Riemann\n" + e.getMessage());
+    }
+    session.transfer(successfulFlowFiles, REL_SUCCESS);
+    session.commit();
```

Comment:
Framework provides this at the conclusion of the onTrigger method as per AbstractProcessor#onTrigger(final ProcessContext context, final ProcessSessionFactory sessionFactory)